### PR TITLE
Optimized wait function and fixed memory leaks

### PR DESCRIPTION
**Pull request by emilwestergren**

### CODE REVIEW

**Comment by emilwestergren** [comment]
Updated.

File: source/base/ThreadPool.ooc [hunk 1]
```diff
@@ -39,13 +46,21 @@ _Task: abstract class {
 		this _mutex lock()
 		if (this _state != _PromiseState Cancelled)
 			this _state = _PromiseState Finished
-		this _mutex unlock()
+		if (this _freeOnCompletion) {
+			this _mutex unlock()
+			this free()
+		}
+		else {
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: `} else {`

File: source/base/ThreadPool.ooc [hunk 2]
```diff
@@ -7,23 +7,30 @@ import os/Time
...
-			}
+	free: override func {
+		this _mutex lock()
+		if (this _state == _PromiseState Unfinished) {
+			this _freeOnCompletion = true
 			this _mutex unlock()
-			Time sleepMilli(_mutexUpdateTime)
 		}
-		status
+		else {
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: `} else {`

**Comment by emilwestergren** [comment]
Updated.

**Pull request merged**